hibernate-core/src/main/java/org/hibernate/persister/entity/AbstractEntityPersister.java

```diff
@@ -4201,6 +4157,42 @@ protected void createLoaders() {
 		);
 	}
```

Comment:
What worries me a bit here is this type of pattern:
```
public class MyEntityPersister extends AbstractEntityPersisters {
    @Override
    protected void createLoaders() {
        super.createLoaders();

        Map loaders = getLoaders();
        // do something with the loaders
    }
}
```
With the new code, the user would only have the 2 eagerly loaded loaders and not the others which might change slightly the behavior.

I don't know if this use case si supported by ORM.

Reply:
It is unfortunately just a line comment, not a javadoc comment, but the intention for `#getLoaders` is clearly just for access during `#createLoaders` - the expectation being that it is a reference usable to add loaders created.  And in fact, that is consistent with all of ORM's use of that method.

In retrospect we maybe should have instead defined `#createLoaders` as passing in the Map and not exposed `#getLoaders`.  But I think its not worth the effort to do that now considering the massive changes here in 6 anyway.

Really, the only supported access to these loaders from a consumer (of the persister) perspective is via `#getAppropriateLoader`.  Long story short, I think we can say that the usage of `#getLoaders` outlined here (really any usage besides putting a created Loader) is not supported.  The method is protected though and the intention comment is not part of the javadoc, so it is conceivable that user code does follow that expectation.  

Personally I say we just disallow that usage ppattern

Reply:
OK, so given that, the patch with the lazy loading would work. Considering how the code is organized, I think we are stuck with using a `Map` and thus a CHM.

@Sanne any insights to share about the concurrency?